File: probe/serial.mjs
/**
 * Serial-port discovery and CTF streaming.
 *
 * Hot-plug friendly: listPorts() is cheap and meant to be polled. Opening a
 * path that later disappears surfaces as an error/close; the daemon stays up.
 */

import { SerialPort } from 'serialport'

/** @typedef {{ path: string, manufacturer: string | null, serialNumber: string | null, vendorId: string | null, productId: string | null, friendlyName: string }} PortInfo */

/**
 * @param {typeof SerialPort} [SerialPortImpl]
 * @returns {Promise<PortInfo[]>}
 */
export async function listPorts(SerialPortImpl = SerialPort) {
  const ports = await SerialPortImpl.list()
  return ports
    .map((p) => ({
      path: p.path,
      manufacturer: p.manufacturer ?? null,
      serialNumber: p.serialNumber ?? null,
      vendorId: p.vendorId ?? null,
      productId: p.productId ?? null,
      friendlyName: friendlyName(p),
    }))
    .sort((a, b) => a.path.localeCompare(b.path))
}

function friendlyName(p) {
  const bits = [p.manufacturer, p.serialNumber].filter(Boolean)
  if (bits.length) return bits.join(' · ')
  if (p.vendorId && p.productId) return `${p.vendorId}:${p.productId}`
  return 'serial port'
}

/**
 * Open a serial port and stream bytes. Sends Zephyr's tracing host command
 * `enable\\r` once open so UART-CTF builds that wait for the host start
 * streaming (harmless when tracing is already on).
 *
 * @param {{
 *   path: string,
 *   baudRate?: number,
 *   onData: (chunk: Buffer) => void,
 *   onError: (err: Error) => void,
 *   onClose: () => void,
 *   SerialPortImpl?: typeof SerialPort,
 *   sendEnable?: boolean,
 * }} opts
 */
export function openSerialStream(opts) {
  const {
    path,
    baudRate = 115200,
    onData,
    onError,
    onClose,
    SerialPortImpl = SerialPort,
    sendEnable = true,
  } = opts

  const port = new SerialPortImpl({ path, baudRate, autoOpen: false })
  let closed = false

  const close = () => {
    if (closed) return
    closed = true
    try {
      if (port.isOpen) port.close(() => {})
      else port.destroy?.()
    } catch {
      /* already gone */
    }
  }

  port.on('data', (chunk) => {
    if (!closed && chunk?.length) onData(Buffer.from(chunk))
  })
  port.on('error', (err) => {
    if (closed) return
    onError(err instanceof Error ? err : new Error(String(err)))
  })
  port.on('close', () => {
    if (closed) return
    closed = true
    onClose()
  })

  port.open((err) => {
    if (err) {
      onError(err instanceof Error ? err : new Error(String(err)))
      return
    }
    if (sendEnable) {
      try {
        port.write(Buffer.from('enable\r'))
      } catch {
        /* best-effort */
      }
    }
  })

  return {
    path,
    baudRate,
    close,
    /** @param {Buffer|Uint8Array|string} data */
    write(data) {
      if (closed || !port.isOpen) return false
      port.write(data)
      return true
    },
  }
}

File: bridge/internal/serial/serial.go
package serial

import (
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/kartben/zephyr-in-the-browser/bridge/internal/protocol"
	goserial "go.bug.st/serial"
)

// Lister can be swapped in tests.
type Lister func() ([]protocol.PortInfo, error)

// Opener opens a port; tests inject fakes.
type Opener func(path string, baud int, onData func([]byte), onError func(error), onClose func()) (Handle, error)

type Handle interface {
	Close() error
	Write([]byte) (int, error)
}

type Port struct {
	mu     sync.Mutex
	handle Handle
	Path   string
	Baud   int
	Phase  string
	Detail string
}

func DefaultLister() Lister {
	return func() ([]protocol.PortInfo, error) {
		ports, err := goserial.GetPortsList()
		if err != nil {
			return nil, err
		}
		sort.Strings(ports)
		out := make([]protocol.PortInfo, 0, len(ports))
		for _, p := range ports {
			name := "serial port"
			out = append(out, protocol.PortInfo{
				Path:         p,
				FriendlyName: name,
			})
		}
		return out, nil
	}
}

func DefaultOpener() Opener {
	return func(path string, baud int, onData func([]byte), onError func(error), onClose func()) (Handle, error) {
		mode := &goserial.Mode{BaudRate: baud}
		port, err := goserial.Open(path, mode)
		if err != nil {
			return nil, err
		}
		h := &liveHandle{port: port, done: make(chan struct{})}
		go func() {
			buf := make([]byte, 4096)
			for {
				n, err := port.Read(buf)
				if n > 0 {
					chunk := make([]byte, n)
					copy(chunk, buf[:n])
					onData(chunk)
				}
				if err != nil {
					select {
					case <-h.done:
						onClose()
					default:
						onError(err)
						onClose()
					}
					return
				}
			}
		}()
		// Zephyr UART-CTF host start (best-effort).
		_, _ = port.Write([]byte("enable\r"))
		return h, nil
	}
}

type liveHandle struct {
	port goserial.Port
	done chan struct{}
	once sync.Once
}

func (h *liveHandle) Close() error {
	h.once.Do(func() { close(h.done) })
	return h.port.Close()
}

func (h *liveHandle) Write(b []byte) (int, error) {
	return h.port.Write(b)
}

func FriendlyName(manufacturer, serial, vid, pid string) string {
	parts := []string{}
	if manufacturer != "" {
		parts = append(parts, manufacturer)
	}
	if serial != "" {
		parts = append(parts, serial)
	}
	if len(parts) > 0 {
		return strings.Join(parts, " · ")
	}
	if vid != "" || pid != "" {
		return fmt.Sprintf("%s:%s", vid, pid)
	}
	return "serial port"
}
